pkg/availability/multisigcollector/internal/parts/certverification/certverification.go

```diff
@@ -4,103 +4,179 @@ import (
 	"errors"
 	"fmt"
 
-	adsl "github.com/filecoin-project/mir/pkg/availability/dsl"
+	"github.com/filecoin-project/mir/pkg/util/sliceutil"
+
+	"github.com/filecoin-project/mir/pkg/availability/multisigcollector/emptycert"
+
+	mscpbtypes "github.com/filecoin-project/mir/pkg/pb/availabilitypb/mscpb/types"
+	apbtypes "github.com/filecoin-project/mir/pkg/pb/availabilitypb/types"
+
 	"github.com/filecoin-project/mir/pkg/availability/multisigcollector/internal/common"
 	"github.com/filecoin-project/mir/pkg/dsl"
-	apb "github.com/filecoin-project/mir/pkg/pb/availabilitypb"
-	"github.com/filecoin-project/mir/pkg/pb/availabilitypb/mscpb"
+	apbdsl "github.com/filecoin-project/mir/pkg/pb/availabilitypb/dsl"
 	t "github.com/filecoin-project/mir/pkg/types"
-	"github.com/filecoin-project/mir/pkg/util/sliceutil"
 )
 
+// State represents the state related to this part of the module.
+type State struct {
+	NextReqID    RequestID
+	RequestState map[RequestID]*RequestState
+}
+
+// RequestID is used to uniquely identify an outgoing request.
+type RequestID = uint64
+
+type RequestState struct {
+	certVerifiedValid map[t.BatchIDString]bool
+}
+
 // IncludeVerificationOfCertificates registers event handlers for processing availabilitypb.VerifyCert events.
 func IncludeVerificationOfCertificates(
 	m dsl.Module,
 	mc *common.ModuleConfig,
 	params *common.ModuleParams,
 	nodeID t.NodeID,
 ) {
+	state := State{
+		NextReqID:    0,
+		RequestState: make(map[RequestID]*RequestState),
+	}
 	// When receive a request to verify a certificate, check that it is structurally correct and verify the signatures.
-	adsl.UponVerifyCert(m, func(cert *apb.Cert, origin *apb.VerifyCertOrigin) error {
-		mscCert, err := verifyCertificateStructure(params, cert)
-		if err != nil {
-			adsl.CertVerified(m, t.ModuleID(origin.Module), err, origin)
+	apbdsl.UponVerifyCert(m, func(cert *apbtypes.Cert, origin *apbtypes.VerifyCertOrigin) error {
+
+		reqID := state.NextReqID
+		state.NextReqID++
+
+		mscCerts, err := verifyCertificateStructure(params, cert)
+		valid, errStr := t.ErrorPb(err)
+
+		if err != nil { // invalid certificate, send cert verified event with valid=False
+			apbdsl.CertVerified(m, origin.Module, valid, errStr, origin)
 			return nil
 		}
 
-		sigMsg := common.SigData(params.InstanceUID, t.BatchID(mscCert.BatchId))
-		dsl.VerifyNodeSigs(m, mc.Crypto,
-			/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
-			/*signatures*/ mscCert.Signatures,
-			/*nodeIDs*/ t.NodeIDSlice(mscCert.Signers),
-			/*context*/ &verifySigsInCertContext{origin},
-		)
+		state.RequestState[reqID] = &RequestState{
+			certVerifiedValid: make(map[t.BatchIDString]bool),
+		}
+		nonEmptyCerts := make([]*mscpbtypes.Cert, 0)
```

Comment:
Do we need to use this and iterate twice? Can we directly issue the signature verification request if the cert is not empty? (And just have a flag for the case that all certificates are empty.)
I think it would simplify the code.

Reply:
done